src/Hl7.Fhir.ElementModel/ScopedNode.cs

```diff
@@ -40,155 +26,31 @@ public ScopedNode(ITypedElement wrapped)
-        public string Id()
-        {
-            if (_cache.Id == null)
-            {
-                _cache.Id = AtResource ? "#" + Current.Children("id").FirstOrDefault()?.Value as string : null;
-            }
-
-            return _cache.Id;
-        }
-
-        public IEnumerable<ScopedNode> ContainedResources()
-        {
-            if (_cache.ContainedResources == null)
-            {
-                _cache.ContainedResources = AtResource ? 
-                    this.Children("contained").Cast<ScopedNode>():
-                    Enumerable.Empty<ScopedNode>();
-            }
-            return _cache.ContainedResources;
-        }
-
-        public class BundledResource
-        {
-            public string FullUrl;
-            public ScopedNode Resource;
-        }
-
-        public IEnumerable<BundledResource> BundledResources()
-        {
-            if (_cache.BundledResources == null)
-            {
-                if (InstanceType == "Bundle")
-                    _cache.BundledResources = from e in this.Children("entry")
-                                              let fullUrl = e.Children("fullUrl").FirstOrDefault()?.Value as string
-                                              let resource = e.Children("resource").FirstOrDefault() as ScopedNode
-                                              select new BundledResource { FullUrl = fullUrl, Resource = resource };
-                else
-                    _cache.BundledResources = Enumerable.Empty<BundledResource>();
-            }
-
-            return _cache.BundledResources;
-        }
-
-        public string FullUrl()
-        {
-            if (_cache.FullUrl == null)
-            {
-                foreach (var parent in ParentResources())
-                {
-                    if (parent.InstanceType == "Bundle")
-                    {
-                        var fullUrl = parent.BundledResources()
-                            .SingleOrDefault(be => this.Location.StartsWith(be.Resource.Location))
-                            ?.FullUrl;
-                        if (fullUrl != null) _cache.FullUrl = fullUrl;
-                    }
-                }
-            }
-
-            return _cache.FullUrl;
-        }
-
-        public IEnumerable<object> Annotations(Type type)
-        {
-            if (type == typeof(ScopedNode))
+            if (type == typeof(ScopedNode) || type == typeof(BaseScopedNode))
                 return new[] { this };
             else
                 return Current.Annotations(type);
         }
 
-        public IEnumerable<ITypedElement> Children(string name = null) =>
-            Current.Children(name).Select(c => new ScopedNode(this, this.ParentResource, c));
+        public override IEnumerable<ITypedElement> Children(string name = null) =>
+            Current.Children(name).Select(c => new ScopedNode(this as ScopedNode, this.ParentResource, c));
+    }
+
+
+    public interface IBundledResource
```

Comment:
Did we have this interface before? Why is this introduced?

Reply:
ScopedNode had a public nested class BundledResource, which was moved now into BaseScopedNode. I think when I tried first to use an interface instead of the abstract class I needed the IBundledResource. now that we are using the BaseScopedNode abstract class, the IBundledResourcecan interface can be removed for the moment.